### ISSUE TITLE
Most AVX/AVX2 functions not exposed

### ISSUE THREAD

**Issue by jackmott** [gnzlbg/sleef-sys#16]
Is this because they make use of `long double` ?
What work needs to be done to get those added? Happy to help.

**Comment by gnzlbg** [comment]
Yes, but there is no way to use `long double` from Rust, so I don't see that anything can be done to expose these.

**Comment by gnzlbg** [comment]
cc @joshtriplett `long double` on FFI is another feature that the FFI WG should at some point tackle. It's impossible to call functions with `long double` because the "call ABI" of the type cannot be emulated.

**Comment by jackmott** [comment]
Gothca, maybe I Will stroll over to the FFI WG and see if I can help there. In the mean time, I can use the sse4 functions twice, I suppose!

**Comment by gnzlbg** [comment]
We can also write a C wrapper that uses double and converts from double to
long double or similar.

**Comment by greatest-ape** [comment]
[Modifying sleef-sys slightly](https://github.com/greatest-ape/sleef-sys/commit/c7f99bd4c07d21bc3b2eb67aaff661547dafcf9d) enables using, at the very least, [this sleef avx function](https://github.com/greatest-ape/simdeez/blob/octasine/src/avx/avx.rs#L723) (coincidentally in a fork of @jackmott's crate). I don't know exactly how this relates the long double problem, but I thought I'd mention it anyway.

**Comment by greatest-ape** [comment]
I put it behind a feature flag here: https://github.com/greatest-ape/sleef-sys/commit/df02548362a4b087e352143da63a523da376bce0

**Comment by greatest-ape** [comment]
@jackmott  Only two functions actually take long double arguments: [Sleef_sincospil_u05 and Sleef_sincospil_u35](https://sleef.org/purec.xhtml) (except for some in the DFT part of sleef which isn't currently compiled). I suspect the reason you're not seeing many AVX/AVX2 functions is that they are only compiled when avx is enabled as a target feature when compiling sleef-sys. #21 should fix this issue, if @gnzlbg is interested in merging it.

**Comment by greatest-ape** [comment]
I'm sorry about the multitude of comments, but for future reference/other readers, I suspect the ABI issue is discussed in:
- https://github.com/rust-lang/rust/issues/53346
- https://github.com/rust-lang/rust/issues/63068
- https://github.com/rust-lang/rfcs/blob/master/text/2574-simd-ffi.md

Progress seems to have stalled on https://github.com/rust-lang/rust/pull/59238, which would implement RFC2574.

In my understanding, my pull request #21 works at some platforms when used carefully, but soundness issues would occur when let's say a function taking a 256 bit wide vector is called on a CPU only supporting 128 bit wide vectors. A fix might be to somehow add

> #[target_feature(enable = X)] in C FFI functions using SIMD types

(to quote the RFC) to all relevant functions in sleef-sys, e.g. X=avx for SIMD type __m256.
